WIP: XOR method for DataQualityFlag and DataQualityDict

tjma12 opened the pull request:
Added XOR method to DataQualityFlag and DataQualityDict classes.

pep8speaks commented:
Hello @tjma12! Thanks for submitting the PR.

 - In the file [`gwpy/segments/flag.py`](https://github.com/gwpy/gwpy/blob/12f2671e13155f113fe20fa683b2556f60277753/gwpy/segments/flag.py), following are the PEP8 issues :

> [Line 1476:80](https://github.com/gwpy/gwpy/blob/12f2671e13155f113fe20fa683b2556f60277753/gwpy/segments/flag.py#L1476): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

duncanmmacleod commented:
@tjma12, I added `WIP` to the PR title, please remove that and ping me here when you are ready for a proper review.

Thanks for your contribution!

tjma12 commented:
It looks like the `copy()` method for `DataQualityDict` isn't working as intended. If I make a `DataQualityDict`, copy it, and modify the underlying segment lists, both dictionaries are modified.

Example from ipython below:
```
In [1]: from gwpy.segments import DataQualityFlag, DataQualityDict, Segment, SegmentList

In [2]: seg1 = Segment(5,10)

In [3]: seg2 = Segment(12,15)

In [4]: seg3 = Segment(4,8)

In [5]: seg4 = Segment(11,14)

In [6]: seglist1 = SegmentList([seg1,seg2])

In [7]: seglist2 = SegmentList([seg3,seg4])

In [8]: flag1 = DataQualityFlag('TEST', active=seglist1, known=seglist1)

In [9]: flag2 = DataQualityFlag('TEST', active=seglist2, known=seglist2)

In [10]: dict1 = DataQualityDict({'f1': flag1})

In [11]: dict2 = DataQualityDict({'f1': flag2})

In [12]: cdict1 = dict1.copy()

In [13]: dict1
Out[13]: 
DataQualityDict([('f1', <DataQualityFlag('TEST',
                 known=[[5 ... 10)
                        [12 ... 15)],
                 active=[[5 ... 10)
                         [12 ... 15)],
                 description=None)>)])

In [14]: cdict1
Out[14]: 
DataQualityDict([('f1', <DataQualityFlag('TEST',
                 known=[[5 ... 10)
                        [12 ... 15)],
                 active=[[5 ... 10)
                         [12 ... 15)],
                 description=None)>)])

In [15]: flag1
Out[15]: 
<DataQualityFlag('TEST',
                 known=[[5 ... 10)
                        [12 ... 15)],
                 active=[[5 ... 10)
                         [12 ... 15)],
                 description=None)>

In [16]: flag2
Out[16]: 
<DataQualityFlag('TEST',
                 known=[[4 ... 8)
                        [11 ... 14)],
                 active=[[4 ... 8)
                         [11 ... 14)],
                 description=None)>

In [17]: flag1 ^= flag2

In [18]: flag1
Out[18]: 
<DataQualityFlag('TEST',
                 known=[[4 ... 5)
                        [8 ... 10)
                        [11 ... 12)
                        [14 ... 15)],
                 active=[[4 ... 5)
                         [8 ... 10)
                         [11 ... 12)
                         [14 ... 15)],
                 description=None)>

In [19]: dict1
Out[19]: 
DataQualityDict([('f1', <DataQualityFlag('TEST',
                 known=[[4 ... 5)
                        [8 ... 10)
                        [11 ... 12)
                        [14 ... 15)],
                 active=[[4 ... 5)
                         [8 ... 10)
                         [11 ... 12)
                         [14 ... 15)],
                 description=None)>)])

In [20]: cdict1
Out[20]: 
DataQualityDict([('f1', <DataQualityFlag('TEST',
                 known=[[4 ... 5)
                        [8 ... 10)
                        [11 ... 12)
                        [14 ... 15)],
                 active=[[4 ... 5)
                         [8 ... 10)
                         [11 ... 12)
                         [14 ... 15)],
                 description=None)>)])
```

duncanmmacleod commented:
@tjma12, can you post that last snippet as a [new issue](/gwpy/gwpy/issues/new)?

tjma12 commented:
@duncanmmacleod I added it as issue #710

duncanmmacleod commented:
@tjma12, given the discussion in #710, what's left for this PR? It would be great to have the following:

- unit tests, probably added to `gwpy/tests/test_segments.py::TestDataQualityFlag::test_math`
- documentation, see `docs/segments/index.rst`: '_Combining DataQualityFlags_'

tjma12 commented:
Once #710 is resolved, this PR will need to be modified to invoke `.copy(deep=True)` and so will the existing `__and__()`, `__or__()`, and `__sub__()` methods that are supposed to return new copies of modified dictionaries rather than in-place modifications. I can add unit tests and documentation as well.

tjma12 commented:
@duncanmmacleod I modified the `DataQualityDict.copy()` method to perform a `deepcopy` if a boolean argument `deep=True` is provided. It defaults to `deep=False`, so any existing code shouldn't be affected. I also modified the not-in-place operations for the `DataQualityDict` to use `deep=True`. I wrote up an example in the documentation and added an XOR unit test.

duncanmmacleod commented on gwpy/segments/flag.py:
```diff
@@ -1569,6 +1581,25 @@ def populate(self, source=DEFAULT_SEGMENT_SERVER,
+        if deep:
+            return deepcopy(self)
+        else:
+            return shallowcopy(self)
```
@tjma12, I would prefer to keep the default open using `super`, just in case the `OrderedDict` does something more than just `shallowcopy`; also, you don't need the `else` here:

```python
        if deep:
            return deepcopy(self)
        return super(DataQualityDict, self).copy()
```

duncanmmacleod commented on gwpy/segments/flag.py:
```diff
@@ -1569,6 +1581,25 @@ def populate(self, source=DEFAULT_SEGMENT_SERVER,
+        Parameters
+        ----------
+        deep : `bool`, optional, default: `False`
+            perform a deep copy of the original dictionary with a fresh 
```
@tjma12, codeclimate is complaining about trailing whitespace on this line, please take a look.

duncanmmacleod commented:
Thanks @tjma12!
